### ISSUE TITLE
Overlapping multi-word phrases break text representation

### ISSUE THREAD

**Issue by savkov** [explosion/spaCy#242]
When two bigrams from the gazetteer overlap in the text, the text representation of the first one is broken. Consider the following code:

```python
import spacy.en
import spacy.matcher
from spacy.attrs import ORTH, TAG, LOWER, IS_ALPHA, FLAG63

patterns = [
    [{LOWER: 'food'}, {LOWER: 'safety'}],
    [{LOWER: 'safety'}, {LOWER: 'standards'}],
]

nlp = spacy.en.English(tagger=False, parser=False, load_vectors=False)

nlp.matcher.add('FOOD', 'FOOD', {}, patterns)

docs = nlp('There are different food safety standards in different countries.')

for e in docs.ents:
    print(e.text, e.label_)
```
Output:
```
>>> food FOOD
>>> safety standards FOOD
```

**Comment by honnibal** [comment]
Thanks! I'll take a look at this next week.

**Comment by syllog1sm** [comment]
Sorry for leaving this to sit for so long.

There's really a deeper design problem here that I should've seen sooner. The code is all doing the right thing.

The problem is that the `Matcher` class automatically sets named entity annotations, but the named entity annotations don't allow overlapping entities. The lack of overlapping entities is not ideal, but it's a standard limitation, and not so problematic. But the `Matcher` class is useful for lots of other situations --- it's a generic way to recognise contiguous sequences of tokens.

I think the `Matcher` class shouldn't set entity annotations. It should just return `Span` objects.

**Comment by savkov** [comment]
I agree. In fact, I am using it at the moment for a few separate groups of annotations, which have to be loaded in separate objects because the overlap. It would be nice to make the overlap optional though because picking the longest match is sometimes desirable. Thoughts?

**Comment by honnibal** [comment]
This is finally fixed in 1.0 — the matcher returns a sequence of `(ent_id, label_id, start, end)` tuples, and you can associate an `on_match` and/or `acceptor` callback with each entity ID. The `on_match` callback is only called after all matches are collected. The callback takes 4 arguments: the matcher, the doc, the index of the match to handle, and the list of matches. This allows the user to resolve the matches considering the total set of matches.
